File: src/core/viewport/build-viewport.ts
// ── Viewport Builder ────────────────────────────────────────────────
//
// Constructs the player-safe viewport from engine state.
// This is the SINGLE function that enforces the information boundary.
// Both the Svelte UI and the CLI harness call this same function.

import { type Seat, type Call, type Hand, type Deal, type Auction, type Contract, type PlayedCard, type Trick, type Card, type Suit } from "../../engine/types";
import type { AuctionContext } from "../contracts/committed-step";
import type { BidAction } from "../contracts/bid-action";
import { evaluateHand } from "../../engine/hand-evaluator";
import { formatCall } from "../display/format";
import { formatHandSummary } from "../display/hand-summary";
import type { BidHistoryEntry, BidResult } from "../contracts/bidding";
import type { BidMeaning } from "../contracts/meaning";
import { isAlertable as isAlertableFromIntent } from "../contracts/alert";
import type { TeachingProjection } from "../contracts/teaching-projection";
import type { TeachingResolution } from "../contracts/teaching-grading";
import type { BidGrade } from "../contracts/teaching-grading";
import type { PracticalScoreBreakdown } from "../contracts/recommendation";
// Minimal interface matching BidFeedback from stores/bidding.svelte.ts.
// Defined here to avoid importing the Svelte store in CLI context.
interface BidFeedbackLike {
  readonly grade: BidGrade;
  readonly userCall: Call;
  readonly expectedResult: BidResult;
  readonly teachingResolution: TeachingResolution;
  readonly practicalRecommendation: { topCandidateBidName: string; topCandidateCall: Call; topScore: number; rationale: string } | null;
  readonly teachingProjection: TeachingProjection | null;
  /** Score breakdown from the practical scorer (when available). */
  readonly practicalScoreBreakdown: PracticalScoreBreakdown | null;
  /** Viewport-safe observation history (projected from AuctionContext). */
  readonly observationHistory?: readonly ObservationStepView[];
}

import type {
  BiddingViewport,
  HandEvaluationView,
  AuctionEntryView,
  BiddingOptionView,
  ViewportBidFeedback,
  ViewportBidGrade,
  ConditionView,
  AlternativeView,
  NearMissView,
  ConventionView,
  TeachingDetail,
  ObservationStepView,
  ObservationView,
  DeclarerPromptViewport,
  PlayingViewport,
  ExplanationViewport,
} from "./player-viewport";

// ── Build Bidding Viewport ──────────────────────────────────────────

export interface BuildBiddingViewportInput {
  readonly deal: Deal;
  readonly userSeat: Seat;
  readonly auction: Auction;
  readonly bidHistory: readonly BidHistoryEntry[];
  readonly legalCalls: readonly Call[];
  readonly faceUpSeats: ReadonlySet<Seat>;
  readonly conventionName: string;
  readonly isUserTurn: boolean;
  readonly currentBidder: Seat;
  /** Active meaning surfaces at the current state (system-card knowledge). */
  readonly activeSurfaces?: readonly BidMeaning[];
}

/**
 * Build a BiddingViewport from engine state.
 *
 * This is the information boundary.  The returned viewport contains
 * ONLY what a player in `userSeat` can legitimately see.
 */
export function buildBiddingViewport(input: BuildBiddingViewportInput): BiddingViewport {
  const {
    deal, userSeat, auction, bidHistory, legalCalls,
    faceUpSeats, conventionName, isUserTurn, currentBidder,
    activeSurfaces,
  } = input;

  // Player's hand only
  const hand = deal.hands[userSeat];
  const eval_ = evaluateHand(hand);
  const handEvaluation: HandEvaluationView = {
    hcp: eval_.hcp,
    shape: eval_.shape,
    isBalanced: eval_.distribution.total === 0, // Shortness 0 = balanced heuristic
    totalPoints: eval_.totalPoints,
    distributionPoints: eval_.distribution,
  };

  // Visible hands — only face-up seats
  const visibleHands: Partial<Record<Seat, Hand>> = {};
  for (const seat of faceUpSeats) {
    visibleHands[seat] = deal.hands[seat];
  }

  // Auction entries with alert annotations
  const auctionEntries: AuctionEntryView[] = [];
  for (let i = 0; i < auction.entries.length; i++) {
    const entry = auction.entries[i]!;
    const historyEntry = bidHistory[i];
    auctionEntries.push({
      seat: entry.seat,
      call: entry.call,
      callDisplay: formatCall(entry.call),
      alertLabel: historyEntry?.alertLabel,
      annotationType: historyEntry?.annotationType,
    });
  }

  // Bidding options from active surfaces (system-card knowledge)
  const biddingOptions: BiddingOptionView[] = [];
  if (activeSurfaces) {
    for (const surface of activeSurfaces) {
      const call = surface.encoding.defaultCall;
      biddingOptions.push({
        call,
        callDisplay: formatCall(call),
        teachingLabel: surface.teachingLabel,
        isAlertable: isAlertableFromIntent(surface.sourceIntent.type),
        recommendation: surface.ranking.recommendationBand as BiddingOptionView["recommendation"],
      });
    }
  }

  return {
    seat: userSeat,
    conventionName,
    hand,
    handEvaluation,
    handSummary: formatHandSummary(eval_),
    visibleHands,
    auctionEntries,
    dealer: deal.dealer,
    vulnerability: deal.vulnerability,
    legalCalls,
    biddingOptions,
    isUserTurn,
    currentBidder,
  };
}

// ── Build Viewport Feedback ─────────────────────────────────────────

/**
 * Convert engine BidFeedback into a viewport-safe ViewportBidFeedback.
 *
 * Strips internal evaluation data, keeping only what a human player
 * would see in the feedback panel.
 */
export function buildViewportFeedback(feedback: BidFeedbackLike): ViewportBidFeedback {
  const grade = feedback.grade as ViewportBidGrade;
  const requiresRetry = grade === "near-miss" || grade === "incorrect";

  // Correct answer (from expected result)
  const correctCall = feedback.expectedResult.call;
  const correctCallDisplay = formatCall(correctCall);
  const correctBidLabel = feedback.expectedResult.alert?.teachingLabel
    ?? feedback.expectedResult.meaning;
  const correctBidExplanation = feedback.expectedResult.explanation;

  // Conditions from teaching projection
  let conditions: ConditionView[] | undefined;
  if (feedback.teachingProjection?.primaryExplanation) {
    conditions = feedback.teachingProjection.primaryExplanation
      .filter((n) => n.kind === "condition")
      .map((n) => ({
        description: n.content,
        passed: n.passed ?? true,
      }));
  }

  // Acceptable alternatives
  let acceptableAlternatives: AlternativeView[] | undefined;
  if (feedback.teachingResolution.acceptableBids?.length) {
    acceptableAlternatives = feedback.teachingResolution.acceptableBids.map((ab) => ({
      call: ab.call,
      callDisplay: formatCall(ab.call),
      label: ab.meaning,
      reason: ab.reason,
      fullCredit: ab.fullCredit,
    }));
  }

  // Near misses
  let nearMisses: NearMissView[] | undefined;
  if (feedback.teachingResolution.nearMissCalls?.length) {
    nearMisses = feedback.teachingResolution.nearMissCalls.map((nm) => ({
      call: nm.call,
      callDisplay: formatCall(nm.call),
      reason: nm.reason,
    }));
  }

  // Partner hand space
  let partnerHandSpace: string | undefined;
  if (feedback.teachingProjection?.handSpace?.partnerSummary) {
    partnerHandSpace = feedback.teachingProjection.handSpace.partnerSummary;
  }

  // Convention contributions
  let conventionsApplied: ConventionView[] | undefined;
  if (feedback.teachingProjection?.conventionsApplied?.length) {
    conventionsApplied = feedback.teachingProjection.conventionsApplied.map((c) => ({
      moduleId: c.moduleId,
      role: c.role,
    }));
  }

  return {
    grade,
    userCall: feedback.userCall,
    userCallDisplay: formatCall(feedback.userCall),
    correctCall,
    correctCallDisplay,
    correctBidLabel,
    correctBidExplanation,
    conditions,
    acceptableAlternatives,
    nearMisses,
    partnerHandSpace,
    conventionsApplied,
    requiresRetry,
  };
}

// ── Build Teaching Detail ───────────────────────────────────────────

/**
 * Extract the teaching detail from a BidFeedback-like object.
 *
 * This is the oracle-side teaching data, available ONLY after grading.
 * Both CLI (--json diagnostics) and UI ("Show Answer" panels) consume
 * this through the same type.
 */
export function buildTeachingDetail(feedback: BidFeedbackLike): TeachingDetail {
  const projection = feedback.teachingProjection;
  const resolution = feedback.teachingResolution;
  const result = feedback.expectedResult;

  return {
    // Correct answer context
    handSummary: result.handSummary,
    fallbackExplanation: result.explanation,

    // Teaching projection (when available)
    primaryExplanation: projection?.primaryExplanation,
    whyNot: projection?.whyNot,
    conventionsApplied: projection?.conventionsApplied,
    meaningViews: projection?.meaningViews,
    callViews: projection?.callViews,

    // Partner hand space
    partnerSummary: projection?.handSpace?.partnerSummary,
    archetypes: projection?.handSpace?.archetypes?.map((a) => ({
      label: a.label,
      hcpRange: a.hcpRange,
      shapePattern: a.shapePattern,
    })),

    // Encoding trace
    encoderKind: projection?.encoderKind,

    // Practical recommendation
    practicalRecommendation: feedback.practicalRecommendation
      ? {
          topCandidateCall: feedback.practicalRecommendation.topCandidateCall,
          rationale: feedback.practicalRecommendation.rationale,
        }
      : undefined,

    // Teaching resolution
    primaryBid: resolution.primaryBid,
    acceptableBids: resolution.acceptableBids?.map((ab) => ({
      call: ab.call,
      meaning: ab.meaning,
      reason: ab.reason,
      fullCredit: ab.fullCredit,
    })),
    nearMissCalls: resolution.nearMissCalls,

    // Decision metadata (from teaching resolution)
    ambiguityScore: resolution.ambiguityScore,
    gradingType: resolution.gradingType,

    // Practical score breakdown
    practicalScoreBreakdown: feedback.practicalScoreBreakdown
      ? { ...feedback.practicalScoreBreakdown }
      : undefined,

    // Evaluation completeness (from teaching projection)
    evaluationExhaustive: projection?.evaluationExhaustive ?? false,
    fallbackReached: projection?.fallbackReached ?? false,

    // Parse tree (from teaching projection)
    parseTree: projection?.parseTree,

    // Observation history (from AuctionContext, viewport-safe projection)
    observationHistory: feedback.observationHistory,
  };
}

// ── Observation log projection ──────────────────────────────────────

/**
 * Project an AuctionContext into a viewport-safe observation history.
 *
 * Strips `resolvedClaim` (moduleId, meaningId, sourceIntent) which are
 * implementation details. Keeps only `publicActions`, `stateAfter`, `actor`,
 * `call`, and `status` — the bridge-observable information.
 */
export function projectObservationHistory(
  ctx: AuctionContext | null | undefined,
): ObservationStepView[] | undefined {
  if (!ctx || ctx.log.length === 0) return undefined;

  return ctx.log.map((step) => ({
    actor: step.actor,
    call: step.call,
    observations: step.publicActions.map(formatObservation),
    kernel: {
      fitAgreed: step.stateAfter.fitAgreed,
      forcing: step.stateAfter.forcing,
      captain: step.stateAfter.captain,
      competition: step.stateAfter.competition,
    },
    status: step.status === "ambiguous" ? "off-system" as const : step.status,
  }));
}

/** Format a BidAction into a human-readable ObservationView. */
function formatObservation(obs: BidAction): ObservationView {
  const parts: string[] = [obs.act];
  if ("feature" in obs && obs.feature) parts.push(obs.feature);
  if ("suit" in obs && obs.suit) parts.push(obs.suit);
  if ("strain" in obs && obs.strain) parts.push(obs.strain);
  if ("targetSuit" in obs && obs.targetSuit) parts.push(obs.targetSuit);
  if ("strength" in obs && obs.strength) parts.push(obs.strength);
  if ("quality" in obs && obs.quality) parts.push(String(obs.quality));
  return { act: obs.act, detail: parts.join("(") + (parts.length > 1 ? ")" : "") };
}

// ── Shared helpers ──────────────────────────────────────────────────

/** Build viewport-safe auction entries from raw auction + bid history. */
function buildAuctionEntries(
  auction: Auction,
  bidHistory: readonly BidHistoryEntry[],
): AuctionEntryView[] {
  const entries: AuctionEntryView[] = [];
  for (let i = 0; i < auction.entries.length; i++) {
    const entry = auction.entries[i]!;
    const historyEntry = bidHistory[i];
    entries.push({
      seat: entry.seat,
      call: entry.call,
      callDisplay: formatCall(entry.call),
      alertLabel: historyEntry?.alertLabel,
      annotationType: historyEntry?.annotationType,
    });
  }
  return entries;
}

/** Filter hands through faceUpSeats, returning only visible ones. */
function filterVisibleHands(
  deal: Deal,
  faceUpSeats: ReadonlySet<Seat>,
): Partial<Record<Seat, Hand>> {
  const visible: Partial<Record<Seat, Hand>> = {};
  for (const seat of faceUpSeats) {
    visible[seat] = deal.hands[seat];
  }
  return visible;
}

// ── Build Declarer Prompt Viewport ──────────────────────────────────

export interface BuildDeclarerPromptViewportInput {
  readonly deal: Deal;
  readonly userSeat: Seat;
  readonly faceUpSeats: ReadonlySet<Seat>;
  readonly auction: Auction;
  readonly bidHistory: readonly BidHistoryEntry[];
  readonly contract: Contract;
  readonly promptMode: "defender" | "south-declarer" | "declarer-swap";
}

/**
 * Build a DeclarerPromptViewport from engine state.
 *
 * Filters hands through faceUpSeats so the component never sees
 * cards the player shouldn't know about.
 */
export function buildDeclarerPromptViewport(
  input: BuildDeclarerPromptViewportInput,
): DeclarerPromptViewport {
  const { deal, userSeat, faceUpSeats, auction, bidHistory, contract, promptMode } = input;

  return {
    userSeat,
    visibleHands: filterVisibleHands(deal, faceUpSeats),
    dealer: deal.dealer,
    vulnerability: deal.vulnerability,
    auctionEntries: buildAuctionEntries(auction, bidHistory),
    contract,
    promptMode,
  };
}

// ── Build Playing Viewport ──────────────────────────────────────────

export interface BuildPlayingViewportInput {
  readonly deal: Deal;
  readonly faceUpSeats: ReadonlySet<Seat>;
  readonly auction?: Auction;
  readonly bidHistory?: readonly BidHistoryEntry[];
  readonly rotated: boolean;
  readonly contract: Contract | null;
  readonly currentPlayer: Seat | null;
  readonly currentTrick: readonly PlayedCard[];
  readonly trumpSuit: Suit | undefined;
  readonly legalPlays: readonly Card[];
  readonly userControlledSeats: readonly Seat[];
  readonly remainingCards: Partial<Record<Seat, readonly Card[]>>;
  readonly tricks: readonly Trick[];
  readonly declarerTricksWon: number;
  readonly defenderTricksWon: number;
}

/**
 * Build a PlayingViewport from engine state.
 *
 * Filters hands through faceUpSeats.  During play, typically the
 * user's hand + dummy are visible.
 */
export function buildPlayingViewport(
  input: BuildPlayingViewportInput,
): PlayingViewport {
  const {
    deal, faceUpSeats, auction, bidHistory, rotated, contract,
    currentPlayer, currentTrick, trumpSuit, legalPlays,
    userControlledSeats, remainingCards, tricks,
    declarerTricksWon, defenderTricksWon,
  } = input;

  return {
    rotated,
    visibleHands: filterVisibleHands(deal, faceUpSeats),
    dealer: deal.dealer,
    vulnerability: deal.vulnerability,
    contract,
    currentPlayer,
    currentTrick,
    trumpSuit,
    legalPlays,
    userControlledSeats,
    remainingCards,
    tricks,
    declarerTricksWon,
    defenderTricksWon,
    auctionEntries: auction ? buildAuctionEntries(auction, bidHistory ?? []) : undefined,
    bidHistory,
  };
}

// ── Build Explanation Viewport ──────────────────────────────────────

export interface BuildExplanationViewportInput {
  readonly deal: Deal;
  readonly userSeat: Seat;
  readonly auction: Auction;
  readonly bidHistory: readonly BidHistoryEntry[];
  readonly contract: Contract | null;
  readonly score: number | null;
  readonly declarerTricksWon: number;
}

/**
 * Build an ExplanationViewport from engine state.
 *
 * All four hands are exposed — this is the review phase where
 * everything is visible.
 */
export function buildExplanationViewport(
  input: BuildExplanationViewportInput,
): ExplanationViewport {
  const { deal, userSeat, auction, bidHistory, contract, score, declarerTricksWon } = input;

  return {
    userSeat,
    allHands: deal.hands,
    dealer: deal.dealer,
    vulnerability: deal.vulnerability,
    auctionEntries: buildAuctionEntries(auction, bidHistory),
    contract,
    score,
    declarerTricksWon,
    bidHistory,
  };
}

File: crates/bridge-session/src/session/build_viewport.rs
//! Viewport builder functions — constructs player-safe viewport DTOs from engine state.
//!
//! This is the SINGLE module that enforces the information boundary. Both the UI
//! (via WASM) and CLI harness call these same functions.
//!
//! Type definitions live in `viewport_types.rs`. This file contains only the
//! builder functions and shared helpers.
//!
//! Ported from TS `src/session/build-viewport.ts`.

use std::collections::{HashMap, HashSet};

use bridge_engine::types::{Auction, BidSuit, Call, Deal, Hand, Seat};
use bridge_engine::{evaluate_hand_hcp, is_balanced};

use super::viewport_types::{
    AuctionEntryView, BidHistoryEntryView, BiddingViewport, BuildBiddingViewportInput,
    BuildDeclarerPromptViewportInput, BuildExplanationViewportInput, BuildPlayingViewportInput,
    DeclarerPromptViewport, ExplanationViewport, HandEvaluationView, PlayingViewport,
};

// ── Format helpers ────────────────────────────────────────────────

/// Format a Call for display: "1\u{2663}", "1NT", "Pass", "Dbl", "Rdbl", "2\u{2665}", etc.
/// Uses Unicode suit symbols to match the TS `formatCall` output.
pub fn format_call(call: &Call) -> String {
    match call {
        Call::Pass => "Pass".to_string(),
        Call::Double => "Dbl".to_string(),
        Call::Redouble => "Rdbl".to_string(),
        Call::Bid { level, strain } => {
            let suit_str = match strain {
                BidSuit::Clubs => "\u{2663}",
                BidSuit::Diamonds => "\u{2666}",
                BidSuit::Hearts => "\u{2665}",
                BidSuit::Spades => "\u{2660}",
                BidSuit::NoTrump => "NT",
            };
            format!("{}{}", level, suit_str)
        }
    }
}

// ── Shared helpers ────────────────────────────────────────────────

/// Build viewport-safe auction entries from raw auction + optional bid history.
pub fn build_auction_entries(
    auction: &Auction,
    bid_history: &[BidHistoryEntryView],
) -> Vec<AuctionEntryView> {
    auction
        .entries
        .iter()
        .enumerate()
        .map(|(i, entry)| {
            let history_entry = bid_history.get(i);
            AuctionEntryView {
                seat: entry.seat,
                call: entry.call.clone(),
                call_display: format_call(&entry.call),
                alert_label: history_entry.and_then(|h| h.alert_label.clone()),
                annotation_type: history_entry.and_then(|h| h.annotation_type),
                meaning: history_entry.and_then(|h| h.meaning.clone()),
            }
        })
        .collect()
}

/// Filter hands through face_up_seats, returning only visible ones.
pub fn filter_visible_hands(deal: &Deal, face_up_seats: &HashSet<Seat>) -> HashMap<Seat, Hand> {
    let mut visible = HashMap::new();
    for &seat in face_up_seats {
        if let Some(hand) = deal.hands.get(&seat) {
            visible.insert(seat, hand.clone());
        }
    }
    visible
}

// ── Builder functions ─────────────────────────────────────────────

/// Build a BiddingViewport from engine state.
///
/// This is the information boundary. The returned viewport contains
/// ONLY what a player in `user_seat` can legitimately see.
pub fn build_bidding_viewport(input: BuildBiddingViewportInput) -> BiddingViewport {
    let hand = input
        .deal
        .hands
        .get(&input.user_seat)
        .cloned()
        .unwrap_or_else(|| Hand { cards: vec![] });

    let eval = evaluate_hand_hcp(&hand);
    let hand_evaluation = HandEvaluationView {
        hcp: eval.hcp,
        shape: eval.shape,
        is_balanced: is_balanced(&eval.shape),
        distribution_points: eval.distribution,
    };

    let hand_summary = format!(
        "{}-{}-{}-{}, {} HCP",
        eval.shape[0], eval.shape[1], eval.shape[2], eval.shape[3], eval.hcp
    );

    let visible_hands = filter_visible_hands(input.deal, input.face_up_seats);
    let auction_entries = build_auction_entries(input.auction, input.bid_history);

    BiddingViewport {
        seat: input.user_seat,
        convention_name: input.convention_name,
        hand,
        hand_evaluation,
        hand_summary,
        visible_hands,
        auction_entries,
        dealer: input.deal.dealer,
        vulnerability: input.deal.vulnerability,
        legal_calls: input.legal_calls.to_vec(),
        is_user_turn: input.is_user_turn,
        current_bidder: input.current_bidder,
        practice_mode: input.practice_mode,
        bid_context: input.bid_context,
        bidding_options: input.bidding_options,
    }
}

/// Build a DeclarerPromptViewport from engine state.
///
/// Filters hands through face_up_seats so the component never sees
/// cards the player shouldn't know about.
pub fn build_declarer_prompt_viewport(
    input: BuildDeclarerPromptViewportInput,
) -> DeclarerPromptViewport {
    DeclarerPromptViewport {
        user_seat: input.user_seat,
        visible_hands: filter_visible_hands(input.deal, input.face_up_seats),
        dealer: input.deal.dealer,
        vulnerability: input.deal.vulnerability,
        auction_entries: build_auction_entries(input.auction, input.bid_history),
        contract: input.contract,
        prompt_mode: input.prompt_mode,
    }
}

/// Build a PlayingViewport from engine state.
///
/// Filters hands through face_up_seats. During play, typically the
/// user's hand + dummy are visible.
pub fn build_playing_viewport(input: BuildPlayingViewportInput) -> PlayingViewport {
    let visible_hands = filter_visible_hands(input.deal, input.face_up_seats);
    let auction_entries = match (input.auction, input.bid_history) {
        (Some(auction), Some(history)) => Some(build_auction_entries(auction, history)),
        (Some(auction), None) => Some(build_auction_entries(auction, &[])),
        _ => None,
    };
    let bid_history = input.bid_history.map(|h| h.to_vec());

    PlayingViewport {
        user_seat: input.user_seat,
        rotated: input.rotated,
        visible_hands,
        dealer: input.deal.dealer,
        vulnerability: input.deal.vulnerability,
        contract: input.contract,
        current_player: input.current_player,
        current_trick: input.current_trick,
        trump_suit: input.trump_suit,
        legal_plays: input.legal_plays,
        user_controlled_seats: input.user_controlled_seats,
        remaining_cards: input.remaining_cards,
        tricks: input.tricks,
        declarer_tricks_won: input.declarer_tricks_won,
        defender_tricks_won: input.defender_tricks_won,
        auction_entries,
        bid_history,
    }
}

/// Build an ExplanationViewport from engine state.
///
/// All four hands are exposed — this is the review phase where
/// everything is visible.
pub fn build_explanation_viewport(input: BuildExplanationViewportInput) -> ExplanationViewport {
    let auction_entries = build_auction_entries(input.auction, &input.bid_history);

    ExplanationViewport {
        user_seat: input.user_seat,
        all_hands: input.deal.hands.clone(),
        dealer: input.deal.dealer,
        vulnerability: input.deal.vulnerability,
        auction_entries,
        contract: input.contract,
        score: input.score,
        declarer_tricks_won: input.declarer_tricks_won,
        defender_tricks_won: input.defender_tricks_won,
        bid_history: input.bid_history,
        tricks: input.tricks,
        play_recommendations: input.play_recommendations,
    }
}

#[cfg(test)]
mod tests {
    use super::super::viewport_types::AnnotationType;
    use super::*;
    use crate::types::PromptMode;
    use bridge_engine::types::{AuctionEntry, Card, Contract, Hand, Rank, Suit, Vulnerability};

    fn make_hand_13() -> Hand {
        Hand {
            cards: vec![
                Card {
                    suit: Suit::Spades,
                    rank: Rank::Ace,
                },
                Card {
                    suit: Suit::Spades,
                    rank: Rank::King,
                },
                Card {
                    suit: Suit::Spades,
                    rank: Rank::Queen,
                },
                Card {
                    suit: Suit::Spades,
                    rank: Rank::Jack,
                },
                Card {
                    suit: Suit::Hearts,
                    rank: Rank::Ace,
                },
                Card {
                    suit: Suit::Hearts,
                    rank: Rank::King,
                },
                Card {
                    suit: Suit::Hearts,
                    rank: Rank::Queen,
                },
                Card {
                    suit: Suit::Diamonds,
                    rank: Rank::Ace,
                },
                Card {
                    suit: Suit::Diamonds,
                    rank: Rank::King,
                },
                Card {
                    suit: Suit::Diamonds,
                    rank: Rank::Queen,
                },
                Card {
                    suit: Suit::Clubs,
                    rank: Rank::Ace,
                },
                Card {
                    suit: Suit::Clubs,
                    rank: Rank::King,
                },
                Card {
                    suit: Suit::Clubs,
                    rank: Rank::Queen,
                },
            ],
        }
    }

    fn make_deal() -> Deal {
        let mut hands = HashMap::new();
        hands.insert(Seat::South, make_hand_13());
        hands.insert(Seat::North, Hand { cards: vec![] });
        hands.insert(Seat::East, Hand { cards: vec![] });
        hands.insert(Seat::West, Hand { cards: vec![] });
        Deal {
            hands,
            dealer: Seat::North,
            vulnerability: Vulnerability::None,
        }
    }

    // ── format_call tests ─────────────────────────────────────────

    #[test]
    fn format_call_pass() {
        assert_eq!(format_call(&Call::Pass), "Pass");
    }

    #[test]
    fn format_call_double() {
        assert_eq!(format_call(&Call::Double), "Dbl");
    }

    #[test]
    fn format_call_redouble() {
        assert_eq!(format_call(&Call::Redouble), "Rdbl");
    }

    #[test]
    fn format_call_bids() {
        assert_eq!(
            format_call(&Call::Bid {
                level: 1,
                strain: BidSuit::Clubs
            }),
            "1\u{2663}"
        );
        assert_eq!(
            format_call(&Call::Bid {
                level: 1,
                strain: BidSuit::NoTrump
            }),
            "1NT"
        );
        assert_eq!(
            format_call(&Call::Bid {
                level: 3,
                strain: BidSuit::Hearts
            }),
            "3\u{2665}"
        );
        assert_eq!(
            format_call(&Call::Bid {
                level: 7,
                strain: BidSuit::Spades
            }),
            "7\u{2660}"
        );
        assert_eq!(
            format_call(&Call::Bid {
                level: 2,
                strain: BidSuit::Diamonds
            }),
            "2\u{2666}"
        );
    }

    // ── Bidding viewport tests ────────────────────────────────────

    #[test]
    fn bidding_viewport_only_shows_user_hand() {
        let deal = make_deal();
        let mut face_up = HashSet::new();
        face_up.insert(Seat::South);

        let viewport = build_bidding_viewport(BuildBiddingViewportInput {
            deal: &deal,
            user_seat: Seat::South,
            auction: &Auction {
                entries: vec![],
                is_complete: false,
            },
            bid_history: &[],
            legal_calls: &[Call::Pass],
            face_up_seats: &face_up,
            convention_name: "Test".to_string(),
            is_user_turn: true,
            current_bidder: Seat::South,
            practice_mode: None,
            bid_context: None,
            bidding_options: None,
        });

        // Only South's hand is visible
        assert_eq!(viewport.visible_hands.len(), 1);
        assert!(viewport.visible_hands.contains_key(&Seat::South));
        assert!(!viewport.visible_hands.contains_key(&Seat::North));
        assert_eq!(viewport.hand.cards.len(), 13);
        assert_eq!(viewport.seat, Seat::South);
        assert!(viewport.is_user_turn);
    }

    #[test]
    fn bidding_viewport_hand_evaluation() {
        let deal = make_deal();
        let face_up = HashSet::new();

        let viewport = build_bidding_viewport(BuildBiddingViewportInput {
            deal: &deal,
            user_seat: Seat::South,
            auction: &Auction {
                entries: vec![],
                is_complete: false,
            },
            bid_history: &[],
            legal_calls: &[],
            face_up_seats: &face_up,
            convention_name: "Test".to_string(),
            is_user_turn: false,
            current_bidder: Seat::North,
            practice_mode: None,
            bid_context: None,
            bidding_options: None,
        });

        // 4-3-3-3 shape with all face cards
        assert_eq!(viewport.hand_evaluation.shape, [4, 3, 3, 3]);
        assert!(viewport.hand_evaluation.is_balanced);
        assert!(viewport.hand_evaluation.hcp > 0);
    }

    // ── Explanation viewport shows all hands ──────────────────────

    #[test]
    fn explanation_viewport_shows_all_hands() {
        let deal = make_deal();

        let viewport = build_explanation_viewport(BuildExplanationViewportInput {
            deal: &deal,
            user_seat: Seat::South,
            auction: &Auction {
                entries: vec![],
                is_complete: false,
            },
            bid_history: vec![],
            contract: None,
            score: None,
            declarer_tricks_won: 0,
            defender_tricks_won: 0,
            tricks: vec![],
            play_recommendations: vec![],
        });

        // All four hands visible in review
        assert_eq!(viewport.all_hands.len(), 4);
        assert!(viewport.all_hands.contains_key(&Seat::North));
        assert!(viewport.all_hands.contains_key(&Seat::East));
        assert!(viewport.all_hands.contains_key(&Seat::South));
        assert!(viewport.all_hands.contains_key(&Seat::West));
    }

    // ── Playing viewport with dummy visible ───────────────────────

    #[test]
    fn playing_viewport_shows_user_and_dummy() {
        let deal = make_deal();
        let mut face_up = HashSet::new();
        face_up.insert(Seat::South);
        face_up.insert(Seat::North); // dummy

        let viewport = build_playing_viewport(BuildPlayingViewportInput {
            deal: &deal,
            user_seat: Seat::South,
            face_up_seats: &face_up,
            auction: None,
            bid_history: None,
            rotated: false,
            contract: Some(Contract {
                level: 3,
                strain: BidSuit::NoTrump,
                doubled: false,
                redoubled: false,
                declarer: Seat::South,
            }),
            current_player: Some(Seat::West),
            current_trick: vec![],
            trump_suit: None,
            legal_plays: vec![],
            user_controlled_seats: vec![Seat::South, Seat::North],
            remaining_cards: HashMap::new(),
            tricks: vec![],
            declarer_tricks_won: 0,
            defender_tricks_won: 0,
        });

        // South and North visible, not East/West
        assert_eq!(viewport.visible_hands.len(), 2);
        assert!(viewport.visible_hands.contains_key(&Seat::South));
        assert!(viewport.visible_hands.contains_key(&Seat::North));
        assert!(!viewport.visible_hands.contains_key(&Seat::East));
    }

    // ── Declarer prompt viewport ──────────────────────────────────

    #[test]
    fn declarer_prompt_viewport_filters_hands() {
        let deal = make_deal();
        let mut face_up = HashSet::new();
        face_up.insert(Seat::South);

        let viewport = build_declarer_prompt_viewport(BuildDeclarerPromptViewportInput {
            deal: &deal,
            user_seat: Seat::South,
            face_up_seats: &face_up,
            auction: &Auction {
                entries: vec![],
                is_complete: false,
            },
            bid_history: &[],
            contract: Contract {
                level: 4,
                strain: BidSuit::Hearts,
                doubled: false,
                redoubled: false,
                declarer: Seat::South,
            },
            prompt_mode: PromptMode::SouthDeclarer,
        });

        assert_eq!(viewport.visible_hands.len(), 1);
        assert!(viewport.visible_hands.contains_key(&Seat::South));
        assert_eq!(viewport.contract.level, 4);
        assert_eq!(viewport.prompt_mode, PromptMode::SouthDeclarer);
    }

    // ── filter_visible_hands ──────────────────────────────────────

    #[test]
    fn filter_visible_hands_empty_set() {
        let deal = make_deal();
        let face_up = HashSet::new();
        let visible = filter_visible_hands(&deal, &face_up);
        assert!(visible.is_empty());
    }

    #[test]
    fn filter_visible_hands_all_seats() {
        let deal = make_deal();
        let mut face_up = HashSet::new();
        face_up.insert(Seat::North);
        face_up.insert(Seat::East);
        face_up.insert(Seat::South);
        face_up.insert(Seat::West);
        let visible = filter_visible_hands(&deal, &face_up);
        assert_eq!(visible.len(), 4);
    }

    // ── build_auction_entries ──────────────────────────────────────

    #[test]
    fn build_auction_entries_with_history() {
        let auction = Auction {
            entries: vec![
                AuctionEntry {
                    seat: Seat::North,
                    call: Call::Bid {
                        level: 1,
                        strain: BidSuit::NoTrump,
                    },
                },
                AuctionEntry {
                    seat: Seat::East,
                    call: Call::Pass,
                },
            ],
            is_complete: false,
        };
        let history = vec![BidHistoryEntryView {
            seat: Seat::North,
            call: Call::Bid {
                level: 1,
                strain: BidSuit::NoTrump,
            },
            meaning: Some("15-17 HCP, balanced".to_string()),
            is_user: false,
            is_correct: None,
            grade: None,
            prior_attempts: None,
            arc_label: None,
            alert_label: Some("15-17".to_string()),
            annotation_type: Some(AnnotationType::Announce),
        }];

        let entries = build_auction_entries(&auction, &history);
        assert_eq!(entries.len(), 2);
        assert_eq!(entries[0].call_display, "1NT");
        assert_eq!(entries[0].alert_label, Some("15-17".to_string()));
        assert_eq!(entries[0].annotation_type, Some(AnnotationType::Announce));
        assert_eq!(entries[1].call_display, "Pass");
        assert_eq!(entries[1].alert_label, None); // no history entry for index 1
    }
}
